src/core/lib/iomgr/ev_epollex_linux.cc

```diff
@@ -163,8 +163,8 @@ static void pollable_unref(pollable* p, const grpc_core::DebugLocation& dbg_loc,
 struct grpc_fd {
   grpc_fd(int fd, const char* name, bool track_err)
       : fd(fd), track_err(track_err) {
-    gpr_mu_init(&orphan_mu);
-    gpr_mu_init(&pollable_mu);
+    orphan_mu = new grpc_core::Mutex();
```

Comment:
Why do we need to dynamically allocate these, rather than just declaring them as direct members of `grpc_fd`?

Reply:
This class is a bit special in a way destroying the mutex in `destroy` not in `dtor` so I decided to use `Mutex*` here to keep the semantic.

Reply:
Oh, I see -- it looks like the `grpc_fd` object go onto a freelist to get reused later, and they don't actually get destroyed until global shutdown time.

Is there going to be any performance problem caused by the mutex being allocated separately?  In particular, I'm wondering if that will hurt cache performance.

An alternative here would be to just not destroy the mutex in `destroy()`.  Conceptually, if we're keeping around the `grpc_fd` object for possible reuse, it seems like we should do the same thing for the mutex -- it doesn't make much sense to cache one without the other.  If we change the code to create and destroy the `grpc_fd` objects with new and delete instead of `gpr_malloc()` and `gpr_free()`, then we can declare the mutexes as direct data members and have them share the lifetime of the `grpc_fd` object.  This seems like a cleaner approach overall.

Also, I notice that we're doing a `memset()` on these mutex fields in `invalidate()` (lines 217-218).  Regardless of what we choose to do here, that probably needs to be updated.

Reply:
I changed the grpc_fd to work like more ordinary objects other than being allocated on pooled memory which uses `gpr_malloc` and `gpr_free`. Presumably this was to reduce the pressure of dynamic memory allocation.

Reply:
Huh, it appears that there is a case where `grpc_fd` can be accessed after it's destroyed, which leads to msan error. ([log](https://source.cloud.google.com/results/invocations/131e5a35-7e4b-46a7-b6ed-c3b261b9d054/targets)) So I should keep `destroy` function anyway.

Reply:
Drive by comment - I spent some time looking at this code in https://github.com/grpc/grpc/issues/23536. As I understand the the freelist approach was just done for correctness rather than performance (so that we can safely get epoll events on destroyed fd's), the comment in https://github.com/grpc/grpc/blob/4a90992d033b341c3f55dbeb695820e893084288/src/core/lib/iomgr/ev_epollex_linux.cc#L331 explains more

Reply:
Thanks, Alex! It's a good information to me.